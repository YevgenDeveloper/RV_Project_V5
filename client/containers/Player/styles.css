@import 'client/styles/variables';

.player {
  width: 100%;
  height: calc(100% - $topbar-height);
  border-radius: 6px;
  padding: 10px;
  background-color: #333;
  display: flex;
  flex-flow: column wrap;

  &:first-child {
    margin-right: 5px;
  }

  &:last-child {
    margin-right: 0;
  }
}

.controls {
  display: flex;
  flex-flow: row nowrap;
  justify-content: space-between;
  margin-bottom: 10px;
}

.preview {
  width: 100%;
  flex: 1;
  display: flex;
  overflow: hidden;
}

.image {
  width: 100%;
  height: auto;
  margin: auto;
  display: block;
  border-radius: 4px;
  background-position: 50% 50%;
  background-repeat: no-repeat;
  animation: image-on 0.3s;
}

@keyframes image-on {
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }
}
