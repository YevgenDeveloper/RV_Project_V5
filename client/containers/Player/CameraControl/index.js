// @flow

import React from 'react';
import classNames from 'classnames/bind';

import { cameras } from 'client/constants';
import { Button } from 'client/components';
import styles from './styles.css';

const cx = classNames.bind(styles);

type Props = {
  disabled: boolean,
  camera: string,
  onChangeCamera: string => void,
};

const CameraControl = (props: Props) => (
  <div className={styles.control}>
    <Button
      label="Cam 1"
      onClick={() => props.onChangeCamera(cameras.CAMERA_1)}
      className={cx(styles.button, {
        [styles.on]: !props.disabled && props.camera === cameras.CAMERA_1,
      })}
      disabled={props.disabled}
    />
    <Button
      label="Cam 2"
      onClick={() => props.onChangeCamera(cameras.CAMERA_2)}
      className={cx(styles.button, {
        [styles.on]: !props.disabled && props.camera === cameras.CAMERA_2,
      })}
      disabled={props.disabled}
    />
    <Button
      label="Cam 3"
      onClick={() => props.onChangeCamera(cameras.CAMERA_3)}
      className={cx(styles.button, {
        [styles.on]: !props.disabled && props.camera === cameras.CAMERA_3,
      })}
      disabled={props.disabled}
    />
    <Button
      label="Cam 4"
      onClick={() => props.onChangeCamera(cameras.CAMERA_4)}
      className={cx(styles.button, {
        [styles.on]: !props.disabled && props.camera === cameras.CAMERA_4,
      })}
      disabled={props.disabled}
    />
  </div>
);

export default CameraControl;
