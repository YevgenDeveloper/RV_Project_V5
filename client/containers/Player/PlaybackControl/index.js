// @flow

import React from 'react';
import classNames from 'classnames/bind';

import { playModes } from 'client/constants';
import { Button } from 'client/components';
import styles from './styles.css';

const cx = classNames.bind(styles);

type Props = {
  disabledBack: boolean,
  disabledForward: boolean,
  disabled: boolean,
  playMode: string,
  onClickPlayBack: void => void,
  onClickPlayForward: void => void,
};

const PlaybackControl = (props: Props) => (
  <div className={styles.control}>
    <Button
      label="Play Back"
      onClick={props.onClickPlayBack}
      className={cx(styles.button, styles.playBack, {
        [styles.on]: props.playMode === playModes.BACKWARD,
      })}
      disabled={props.disabledBack || props.disabled}
    />
    <Button
      label="Play Forward"
      onClick={props.onClickPlayForward}
      className={cx(styles.button, styles.playForward, {
        [styles.on]: props.playMode === playModes.FORWARD,
      })}
      disabled={props.disabledForward || props.disabled}
    />
  </div>
);

export default PlaybackControl;
