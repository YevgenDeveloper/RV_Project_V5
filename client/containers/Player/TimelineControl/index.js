// @flow

import React, { Component } from 'react';

import styles from './styles.css';

type Props = {
  currentImage: number,
  totalImages: number,
  disabled: boolean,
  onChangeImage: number => void,
};

type State = {
  totalImages: number,
  pixelsPerFrame: number,
  currentImage: number,
  sliderPosition: number,
};

export default class TimelineControl extends Component<Props, State> {
  state = {
    totalImages: 0,
    pixelsPerFrame: 1,
    // eslint-disable-next-line
    currentImage: 0,
    sliderPosition: 0,
  };

  componentDidMount() {
    document.addEventListener('resize', this.updateSlider);
    this.updateSlider();
  }

  componentDidUpdate(prevProps: Props, prevState: State) {
    if (
      prevProps.currentImage !== this.props.currentImage ||
      prevState.totalImages !== this.state.totalImages
    ) {
      this.updateSlider();
    }
  }

  componentWillUnmount() {
    document.removeEventListener('resize', this.updateSlider);
  }

  static getDerivedStateFromProps(nextProps: Props) {
    return {
      currentImage: nextProps.currentImage,
      totalImages: nextProps.totalImages,
    };
  }

  updateSlider = () => {
    this.setState(state => {
      if (!state.totalImages) {
        return null;
      }

      const pixelsPerFrame = this.ref.offsetWidth / state.totalImages;

      return {
        pixelsPerFrame,
        sliderPosition: state.currentImage * pixelsPerFrame,
      };
    });
  };

  handleClick = (event: MouseEvent) => {
    const { left } = this.ref.getBoundingClientRect();
    const sliderPosition = event.pageX - left;

    this.setState({ sliderPosition }, () => {
      this.props.onChangeImage(
        Math.round(sliderPosition / this.state.pixelsPerFrame),
      );
    });
  };

  ref: HTMLButtonElement;

  render() {
    const { sliderPosition } = this.state;

    return (
      <button
        type="button"
        ref={ref => {
          if (ref) this.ref = ref;
        }}
        className={styles.timeline}
        onClick={this.handleClick}
        disabled={this.props.disabled}
      >
        <div
          className={styles.path}
          style={{ width: `${sliderPosition > 4 ? sliderPosition - 4 : 0}px` }}
        />
        <div
          className={styles.slider}
          style={{ transform: `translateX(${sliderPosition}px)` }}
        />
      </button>
    );
  }
}
