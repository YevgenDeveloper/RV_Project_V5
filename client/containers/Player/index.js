// @flow

import type { PlayerType } from 'client/types';

import React, { Component } from 'react';

import { playModes } from 'client/constants';
import { getImageUrl } from 'client/services';
import { LoaderSpinner } from 'client/components';
import CameraControl from './CameraControl';
import PlaybackControl from './PlaybackControl';
import TimelineControl from './TimelineControl';
import styles from './styles.css';

type Props = {
  isLoading: boolean,
  currentRide: string,
  images: string[],
  player: PlayerType,
  update: ({}) => void,
};

type State = {
  totalImages: number,
  currentImage: number,
};

export default class Player extends Component<Props, State> {
  state = {
    totalImages: 0,
    currentImage: this.props.player ? this.props.player.currentImage : 0,
  };

  componentDidUpdate(prevProps: Props) {
    if (prevProps.player.currentImage !== this.props.player.currentImage) {
      this.handleChangeImage(this.props.player.currentImage);
    }

    if (prevProps.player.playMode !== this.props.player.playMode) {
      this.updateImage();
    }
  }

  static getDerivedStateFromProps(props: Props) {
    return {
      totalImages: props.images ? props.images.length : 0,
    };
  }

  getImageUrl = () => {
    const { player, images, currentRide } = this.props;
    const { currentImage } = this.state;

    if (!images || !images[currentImage]) {
      return undefined;
    }

    return getImageUrl(currentRide, player.camera, images[currentImage]);
  };

  handleChangePlayMode = (playMode: string) => {
    this.props.update({
      playMode:
        playMode === this.props.player.playMode ? playModes.PAUSE : playMode,
    });
  };

  handleChangeImage = (currentImage: number) => {
    this.setState({ currentImage });
    this.props.update({ currentImage });
  };

  updateImage = () => {
    const { playMode } = this.props.player;
    const { currentImage, totalImages } = this.state;

    if (
      playMode === playModes.PAUSE ||
      this.props.isLoading ||
      (playMode === playModes.FORWARD && currentImage === totalImages) ||
      (playMode === playModes.BACKWARD && currentImage === 0)
    ) {
      this.props.update({
        playMode: playModes.PAUSE,
        currentImage,
      });
      return;
    }

    this.setState(state => ({
      currentImage:
        playMode === playModes.FORWARD
          ? state.currentImage + 1
          : state.currentImage - 1,
    }));
  };

  render() {
    const { isLoading, player } = this.props;

    return (
      <div className={styles.player}>
        <div className={styles.controls}>
          <CameraControl
            camera={player.camera}
            onChangeCamera={(camera: string) => this.props.update({ camera })}
            disabled={this.props.isLoading}
          />
          <PlaybackControl
            playMode={player.playMode}
            onClickPlayBack={() =>
              this.handleChangePlayMode(playModes.BACKWARD)
            }
            onClickPlayForward={() =>
              this.handleChangePlayMode(playModes.FORWARD)
            }
            disabledBack={this.state.currentImage === 0}
            disabledForward={this.state.currentImage === this.state.totalImages}
            disabled={this.props.isLoading}
          />
          <TimelineControl
            currentImage={this.state.currentImage}
            totalImages={this.state.totalImages}
            onChangeImage={this.handleChangeImage}
            disabled={this.props.isLoading}
          />
        </div>
        <button
          type="button"
          className={styles.preview}
          onClick={() => this.handleChangePlayMode(playModes.FORWARD)}
          disabled={isLoading}
        >
          {isLoading ? (
            <LoaderSpinner />
          ) : (
            <img
              alt=""
              className={styles.image}
              src={this.getImageUrl()}
              onLoad={this.updateImage}
              onError={() => console.error('Error loading image')}
            />
          )}
        </button>
      </div>
    );
  }
}
