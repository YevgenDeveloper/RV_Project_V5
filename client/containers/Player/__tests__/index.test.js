// @flow

import type { ShallowWrapper } from 'enzyme';

import React from 'react';
import { shallow } from 'enzyme';
import Player from '..';

it('renders correctly', () => {
  const wrapper: ShallowWrapper = shallow(
    <Player
      disabled={false}
      player={{ camera: 'Test', playMode: 'Test', currentImage: 0 }}
    />,
  );

  expect(wrapper).toMatchSnapshot();
});
