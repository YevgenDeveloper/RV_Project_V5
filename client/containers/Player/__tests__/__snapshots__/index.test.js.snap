// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Player
    disabled={false}
    player={
      Object {
        "camera": "Test",
        "currentImage": 0,
        "playMode": "Test",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="controls"
        >
          <CameraControl
            camera="Test"
            onChangeCamera={[Function]}
          />
          <PlaybackControl
            disabledBack={true}
            disabledForward={true}
            onClickPlayBack={[Function]}
            onClickPlayForward={[Function]}
            playMode="Test"
          />
          <TimelineControl
            currentImage={0}
            onChangeImage={[Function]}
            totalImages={0}
          />
        </div>,
        <button
          className="preview"
          onClick={[Function]}
          type="button"
        >
          <img
            alt=""
            className="image"
            onError={[Function]}
            onLoad={[Function]}
          />
        </button>,
      ],
      "className": "player",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <CameraControl
              camera="Test"
              onChangeCamera={[Function]}
            />,
            <PlaybackControl
              disabledBack={true}
              disabledForward={true}
              onClickPlayBack={[Function]}
              onClickPlayForward={[Function]}
              playMode="Test"
            />,
            <TimelineControl
              currentImage={0}
              onChangeImage={[Function]}
              totalImages={0}
            />,
          ],
          "className": "controls",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "camera": "Test",
              "disabled": undefined,
              "onChangeCamera": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "disabled": undefined,
              "disabledBack": true,
              "disabledForward": true,
              "onClickPlayBack": [Function],
              "onClickPlayForward": [Function],
              "playMode": "Test",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "currentImage": 0,
              "disabled": undefined,
              "onChangeImage": [Function],
              "totalImages": 0,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <img
            alt=""
            className="image"
            onError={[Function]}
            onLoad={[Function]}
          />,
          "className": "preview",
          "disabled": undefined,
          "onClick": [Function],
          "type": "button",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "alt": "",
            "className": "image",
            "onError": [Function],
            "onLoad": [Function],
            "src": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="controls"
          >
            <CameraControl
              camera="Test"
              onChangeCamera={[Function]}
            />
            <PlaybackControl
              disabledBack={true}
              disabledForward={true}
              onClickPlayBack={[Function]}
              onClickPlayForward={[Function]}
              playMode="Test"
            />
            <TimelineControl
              currentImage={0}
              onChangeImage={[Function]}
              totalImages={0}
            />
          </div>,
          <button
            className="preview"
            onClick={[Function]}
            type="button"
          >
            <img
              alt=""
              className="image"
              onError={[Function]}
              onLoad={[Function]}
            />
          </button>,
        ],
        "className": "player",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <CameraControl
                camera="Test"
                onChangeCamera={[Function]}
              />,
              <PlaybackControl
                disabledBack={true}
                disabledForward={true}
                onClickPlayBack={[Function]}
                onClickPlayForward={[Function]}
                playMode="Test"
              />,
              <TimelineControl
                currentImage={0}
                onChangeImage={[Function]}
                totalImages={0}
              />,
            ],
            "className": "controls",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "camera": "Test",
                "disabled": undefined,
                "onChangeCamera": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "disabled": undefined,
                "disabledBack": true,
                "disabledForward": true,
                "onClickPlayBack": [Function],
                "onClickPlayForward": [Function],
                "playMode": "Test",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "currentImage": 0,
                "disabled": undefined,
                "onChangeImage": [Function],
                "totalImages": 0,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <img
              alt=""
              className="image"
              onError={[Function]}
              onLoad={[Function]}
            />,
            "className": "preview",
            "disabled": undefined,
            "onClick": [Function],
            "type": "button",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "",
              "className": "image",
              "onError": [Function],
              "onLoad": [Function],
              "src": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
